unittests/fork_db_tests.cpp

```diff
@@ -0,0 +1,131 @@
+#include <eosio/chain/types.hpp>
+#include <eosio/chain/fork_database.hpp>
+#include <eosio/testing/tester.hpp>
+#include <fc/bitutil.hpp>
+#include <boost/test/unit_test.hpp>
+
+
+namespace eosio::chain {
+
+inline block_id_type make_block_id(block_num_type block_num) {
+   block_id_type id = fc::sha256::hash(std::to_string(block_num) + std::to_string(fc::time_point::now().time_since_epoch().count()));
+   id._hash[0] &= 0xffffffff00000000;
+   id._hash[0] += fc::endian_reverse_u32(block_num); // store the block num in the ID, 160 bits is plenty for the hash
+   return id;
+}
+
+struct block_state_accessor {
+
+   static auto make_genesis_block_state() {
+      block_state_ptr root = std::make_shared<block_state>();
+      block_id_type genesis_id = make_block_id(10);
+      root->block_id = genesis_id;
+      root->header.timestamp = block_timestamp_type{10};
+      root->core = finality_core::create_core_for_genesis_block(10);
+      root->best_qc_claim = {.block_num = 10, .is_strong_qc = false};
+      return root;
+   }
+
+   // use block_num > 10
+   static auto make_unique_block_state(block_num_type block_num, const block_state_ptr& prev) {
+      block_state_ptr bsp = std::make_shared<block_state>();
+      bsp->block_id = make_block_id(block_num);
+      bsp->header.timestamp.slot = prev->header.timestamp.slot + 1;
+      bsp->header.previous = prev->id();
+      block_ref parent_block {
+         .block_id  = prev->id(),
+         .timestamp = prev->timestamp()
+      };
+      bsp->core = prev->core.next(parent_block, prev->best_qc_claim);
+      bsp->best_qc_claim = bsp->core.latest_qc_claim();
+      bsp->updated_core = bsp->core.next_metadata(bsp->best_qc_claim);
+      return bsp;
+   }
+
+   static auto get_best_qc_claim(const block_state_ptr& bs) {
+      return bs->best_qc_claim;
+   }
+};
+
+} // namespace eosio::chain
+
+using namespace eosio::chain;
+
+BOOST_AUTO_TEST_SUITE(fork_database_tests)
+
+BOOST_AUTO_TEST_CASE(update_best_qc) try {
+
```

Comment:
Would you explain the high level setup of those blocks (bsps)?

Reply:
Added some comments